### PR TITLE
Refactor runtime to full async/await; make officially embeddable

### PR DESCRIPTION
**Pull request by nolar**
Make operator termination even more graceful (resource-cleanable), the task orchestration — async, and the operator core — embeddable (and officially documented).

> Issue : #142.

## Description

Previously in #147, there was an attempt to rework the asyncio task management with the "should-stop" event and asyncio task cancellations. It made the operator to react to the existing signals, but still had some problems with proper cleanups.

Specifically, when a `aiohttp.ClientSession` is added (a PR based on this one — to be created), there is no place to properly create it and close it, since all the tasks are orchestrated in a synchronous function.

This led to few issues with some `aiohttp` connections not being closed properly, issuing the warnings/prints on the process exiting (and a lot of them in tests).

### Fully async coroutines

With this PR, and operator's body is extracted into a self-contained coroutine (`operator()`), which does everything: task spawning, task orchestration, resource creation and cleanups, etc.

The synchronous functions `run()` and `create_tasks()` are left as wrappers for `operator()` and `spawn_tasks()` respectively. (`create_tasks()` is also deprecated.)

### Graceful termination

With this PR, the operator termination is now even more graceful both in case of normal exit ("should-stop" event set) and in case of error. In the latter case, all tasks & sub-tasks are cancelled without a tolerable graceful timeout, but they are still given time to cleanup their resources (e.g. in `try-finally`).

## Embeddability

These function were already exposed via the public interface (accessible via `kopf.run`/`kopf.create_tasks`). 

With this PR, all the new functions are also exposed, and this is made a documented feature of embeddability — i.e. ability to embed Kopf operators into third-party applications by explicit invocation/orchestration of the Kopf tasks.

_(As an example of potential usage (not in the plans): an application that listens for the resources updates and keeps their state in-memory, while showing that state in a UI/WebUI.)_ 

## Types of Changes

- Bug fix (non-breaking change which fixes an issue)
- Refactor/improvements
- Documentation / non-code

## Review
_List of tasks the reviewer must do to review the PR_
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: docs/embedding.rst [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        thread = threading.Thread(target_fn=kopf_thread)
+        thread.start()
+        # ...
+        thread.join()
+
+
+.. warning::
+    It is not recommended to run Kopf in the same event-loop with other routines
+    or applications: it considers all tasks in the event-loop as spawned by its
+    workers and handlers, and cancells them when it exits.
```
Comment (samurang87) [inline docs/embedding.rst]: "cancels"

File: docs/embedding.rst [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    There are some basic safety measures to not cancel the tasks existed before
```
Comment (samurang87) [inline docs/embedding.rst]: "cancel tasks existing prior to the operator's startup"

File: kopf/reactor/running.py [hunk 3]
```diff
@@ -130,29 +149,104 @@ def create_tasks(
...
+            # If the operator is cancelled, propagate the cancellation to all the sub-tasks.
+            hung_cancelled, hung_left = await _stop(hung_tasks, title="Hung", cancelled=True)
+            raise
+        else:
+            # If the operator is intact, but the timeout is reached, forcely cancel the sub-tasks.
+            hung_cancelled, hung_left = await _stop(hung_pending, title="Hung", cancelled=False)
+
+    # If succeeded or if cancellation is silenced, re-raise from failed tasks (if any).
+    await _reraise(root_done | root_cancelled | hung_done | hung_cancelled)
+    # assert not root_left
```
Comment (samurang87) [inline kopf/reactor/running.py]: are the comments intentional? :)

File: docs/embedding.rst (hunk 1 above)
Comment (nolar) [reply docs/embedding.rst]: Fixed.

File: docs/embedding.rst (hunk 2 above)
Comment (nolar) [reply docs/embedding.rst]: Fixed.

File: kopf/reactor/running.py (hunk 3 above)
Comment (nolar) [reply kopf/reactor/running.py]: Removed.

**Comment by nolar** [comment]
@samurang87 Thanks. All is fixed now.

**Comment by psycho-ir** [comment]
👍

**Pull request merged**